JBMC Java Core models library

dcattaruzza opened the pull request:
We aim at internalizing a portion of the model-library by turning the jar into an include file.
This will enable us to load the corresponding buffer (or any we need henceforth) from an include file in a similar fashion as we load from a jar file.
Currently this only applies to org.cprover.CProver, and later we'll add other java sources.

dcattaruzza committed:
Add constructors for having memory-loaded jar files
This allows the jar_file class to load from a buffer (c array) as opposed to a file

dcattaruzza committed:
Internalize core models of the Java Class Library

The 'core models' is a set of .class files that model core classes from the Java
Class Library, such as java.lang.Object or java.lang.Thread. A minimum is
necessary for CBMC to understand, e.g., when a new thread is created. These core
classes live in `src/java_bytecode/library`.

This commit adds support to compile and pack the core classes into a single jar
file, core-models.jar, and a converter program that transforms that .jar file
into a C-language array of data that can then be linked into CBMC, thus making
the .jar file be present in the data segment of the CBMC ELF.

Other modifications:
- New option --no-core-models, allowing to disable the loading of the internal
  core models
- make and cmake now compile the core models into a single core-models.jar
- Add regression one regression tests ensuring the the core-models.jar is loaded
  by default

tautschnig commented:
What is a "core model"? The commit message of the second commit has a lot of text, but I'm failing to understand it right at the beginning.

tautschnig commented on regression/cbmc-java/coreModels/test.desc:
```diff
@@ -0,0 +1,8 @@
+^SIGNAL=0$
+^Symbol\s*\.*\: java\:\:org\.cprover\.CProver\.\<clinit\>\:\(\)V$
+--
+tests that the core models are  being loaded by checking if the static initializer for the CProver class was
```
This is missing a second `--`, unless you really want the test to check that that text does not appear in the output.

cesaro replied:
done

tautschnig commented on src/java_bytecode/Makefile:
```diff
@@ -36,10 +36,52 @@ INCLUDES= -I ..
+JAR := jar
+JARFLAGS := -cf
+
+.NOTPARALLEL: core-models.jar
```
Why?

dcattaruzza replied:
Java classes compile together, meaning two javac calls in this make may trigger the same .java compile for an auxiliary class. This avoids the race condition

tautschnig commented on src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -206,6 +207,24 @@ bool java_bytecode_languaget::typecheck(
 
+  // Must load java.lang.Object first to avoid stubbing
+  // FIXME: its unclear if this ordering has to be done
+  // here or hard coded inside the classloader.
```
File an issue to track this?

cesaro replied:
@dcattaruzza Unless @tautschnig or @peterschrammel think it's better to move this code to the class loader, I propose to leave this code here and update the comment above to say that "This ordering could alternatively be enforced by moving the code below to the class loader."

dcattaruzza commented:
Core models are java class models that are included in the cbmc executable (ie in the core of the analysis).  Currently all of our models exist in the model library. The idea is to make a subset of these available with open source jbmc, packaged directly in the executable

tautschnig commented on src/java_bytecode/java_class_loader.cpp:
```diff
@@ -74,38 +83,76 @@ void java_class_loadert::set_java_cp_include_files(
 
-java_bytecode_parse_treet &java_class_loadert::get_parse_tree(
+/// Retrieves a class file from a jar and loads it into the tree
+bool java_class_loadert::get_class_file(
```
Why does the class loader need to be rewritten when the commit message says this is about putting stuff in a library file?

dcattaruzza replied:
the class loader needs to load from the internal file buffer

tautschnig reviewed (commented):
I don't claim I need to understand each and every thing, but it really scares me when a commit message sounds innocent (though unclear), but then some core bits of the code are changed as well.

peterschrammel commented on src/java_bytecode/jar_file.cpp:
```diff
@@ -26,6 +23,29 @@ jar_filet::jar_filet(
+/// to a jar file.
+jar_filet::jar_filet(
+  java_class_loader_limitt &limit,
+  const void *pMem,
```
no camel case

tautschnig replied:
Also, it would be nice to use a parameter name that is actually descriptive.

peterschrammel commented on src/java_bytecode/mz_zip_archive.h:
```diff
@@ -24,6 +24,13 @@ class mz_zip_archivet final
+  /// \param pMem pointer to the memory buffer
+  /// \param size size of the buffer
+  /// \throw Throws std::runtime_error if file cannot be opened
+  explicit mz_zip_archivet(const void *pMem, size_t size);
```
remove explicit

peterschrammel commented on src/java_bytecode/mz_zip_archive.cpp:
```diff
@@ -25,6 +25,14 @@ class mz_zip_archive_statet final:public mz_zip_archive
       throw std::runtime_error("MZT: Could not load a file: "+filename);
   }
+
+  explicit mz_zip_archive_statet(const void *pMem, size_t size):
```
remove explicit

peterschrammel commented on src/java_bytecode/mz_zip_archive.cpp:
```diff
@@ -25,6 +25,14 @@ class mz_zip_archive_statet final:public mz_zip_archive
+    mz_zip_archive({ })
+  {
+    if(MZ_TRUE!=mz_zip_reader_init_mem(this, pMem, size, 0))
+      throw std::runtime_error("MZT: Could not load data from memory");
```
Does the user know what `MZT` is?

dcattaruzza replied:
MZT is as existing messages in this file.
Do you want these changed?

cesaro replied:
@dcattaruzza I think it's a good idea to change that message to something more readable, yes. Have a look to the other messages as well.

peterschrammel commented on src/java_bytecode/Makefile:
```diff
@@ -36,10 +36,52 @@ INCLUDES= -I ..
+java_class_loader.o: java_core_models.inc java_class_loader.cpp
+	$(CXX) -c $(CP_CXXFLAGS) -o $@ java_class_loader.cpp
+
+java_core_models.inc: converter core-models.jar
```
This doesn't work on Windows.

cesaro replied:
@dcattaruzza Appveyor is faling, is it due to this? I think it's not. The idea of compiling multiple files into an `.inc` file works for the C library, so it should work for the Java class library.

dcattaruzza replied:
As far as I can tell, Appveyor fails because of shell find -name *.java does not find the java files. 
I'm investigating.

tautschnig replied:
Still you should be using `$(EXEEXT)` here.

peterschrammel commented:
The commits are a bit mixed up, which makes it hard to review. I suggest the following split:
1. refactor existing class loader (to make it easy to choose between loading from external or internal jar file later on)
2. extend class loader to be able to read from internal jar (not using this functionality yet)
3. jar file converter
4. add embedded jar file global and use functionality of 2 to wire everything together (should only require minimal changes to the class loader)

dcattaruzza commented:
@peterschrammel do you mean split the commits or the PR? because I started by doing a smaller PR in the lines of what you're saying and had to include the rest to provide a regression test (the test on the jar buffer is the core models loading) @tautschnig request.

peterschrammel commented:
Yes, split the commits. I suggest to add the end-to-end regression tests as part of 4 or in a separate commit 5.

dcattaruzza committed:
Refactor jar loading

cesaro commented on src/java_bytecode/mz_zip_archive.cpp:
```diff
@@ -41,6 +49,9 @@ static_assert(sizeof(mz_uint)<=sizeof(size_t),
 mz_zip_archivet::mz_zip_archivet(const std::string &filename):
   m_state(new mz_zip_archive_statet(filename)) { }
 
+mz_zip_archivet::mz_zip_archivet(const void *pMem, size_t size):
```
Lower case

cesaro commented on .gitignore:
```diff
@@ -45,6 +45,7 @@ src/ansi-c/gcc_builtin_headers_tm.inc
 src/ansi-c/gcc_builtin_headers_mips.inc
 src/ansi-c/gcc_builtin_headers_power.inc
 src/ansi-c/gcc_builtin_headers_ubsan.inc
+src/java_bytecode/java_core_models.inc
```
Please rephrase the commit message as follows:

```
Internalize core models of the Java Class Library

The 'core models' is a set of .class files that model core classes from the Java
Class Library, such as java.lang.Object or java.lang.Thread. A minimum is
necessary for CBMC to understand, e.g., when a new thread is created. These core
classes live in `src/java_bytecode/library`.

This commit adds support to compile and pack the core classes into a single jar
file, core-models.jar, and a converter program that transforms that .jar file
into a C-language array of data that can then be linked into CBMC, thus making
the .jar file be present in the data segment of the CBMC ELF.

Other modifications:
- New option --no-core-models, allowing to disable the loading of the internal
  core models
- make and cmake now compile the core models into a single core-models.jar
- Add regression one regression tests ensuring the the core-models.jar is loaded
  by default
```

dcattaruzza replied:
done

cesaro commented on src/java_bytecode/CMakeLists.txt:
```diff
@@ -1,5 +1,28 @@
+
+add_library(java_bytecode ${c_sources} ${sources} ${headers} )
+add_dependencies(java_bytecode core_models_files)
+target_sources(java_bytecode PUBLIC ${sources} ${headers})
```
@dcattaruzza I remember running twice `cmake` and it rebuilding for a second time the .jar file, which it should be constructed only once. Has this been fixed?

dcattaruzza replied:
Yes

cesaro commented on src/java_bytecode/java_bytecode_language.h:
```diff
@@ -35,19 +36,20 @@ Author: Daniel Kroening, kroening@kroening.com
-  " --lazy-methods-extra-entry-point METHODNAME\n"                                              \
-  "                                  treat METHODNAME as a possible program entry point for\n"  \
-  "                                  the purpose of lazy method loading\n"                      \
+  " --no-core-models                 don't load the internally provided core models\n"           /* NOLINT(*) */ \
```
Better: "don't load internally provided models for core classes in the Java Class Library"

cesaro reviewed (commented):
Remarks and proposal for new commit message

cesaro reviewed (approved):
Merge when the Appveyor issue is fixed, Travis is passing, and it has been adequately rebased.

tautschnig commented on .gitignore:
```diff
@@ -115,6 +116,7 @@ src/ansi-c/file_converter
 src/ansi-c/file_converter.exe
 src/ansi-c/library/converter
 src/ansi-c/library/converter.exe
+src/java_bytecode/converter
```
What about the .exe variant?

dcattaruzza replied:
done

cesaro commented on src/java_bytecode/Makefile:
```diff
@@ -36,10 +36,53 @@ INCLUDES= -I ..
+
+CLASSPATH := SOURCE_DIR
+
+all_javas := $(wildcard $(CPROVER_SRC_DIR)/*.java) #$(shell find $(SOURCE_DIR) -name '*.java')
```
Change to `$(wildcard library/*.java library/*/*.java library/*/*/*.java library/*/*/*/*.java)`.
Variables in a `Makefile` are always capital letters.

cesaro replied:
Done.

tautschnig commented on src/java_bytecode/Makefile:
```diff
@@ -36,10 +36,52 @@ INCLUDES= -I ..
+	$(CXX) $(CP_CXXFLAGS) -o $@ $<
+
+java_class_loader.o: java_core_models.inc java_class_loader.cpp
+	$(CXX) -c $(CP_CXXFLAGS) -o $@ java_class_loader.cpp
```
Again, `ansi-c/Makefile` may serve as reference. In short: just use `java_class_loader$(OBJEXT): java_core_models.inc` (nothing else required, in particular no commands)

tautschnig replied:
This should really just be `java_class_loader$(OBJEXT): java_core_models.inc`, no `java_class_loader.cpp` or the `$(CXX) ...` rule

tautschnig commented on src/java_bytecode/Makefile:
```diff
@@ -36,10 +36,52 @@ INCLUDES= -I ..
+	$(CXX) -c $(CP_CXXFLAGS) -o $@ java_class_loader.cpp
+
+java_core_models.inc: converter core-models.jar
+	./converter JAVA_CORE_MODELS < core-models.jar > $@
```
... and here.

tautschnig commented on src/java_bytecode/Makefile:
```diff
@@ -36,10 +36,52 @@ INCLUDES= -I ..
+core-models.jar: $(all_classes)
+	$(JAR) $(JARFLAGS) $@ -C $(BINARY_DIR) .
+
+CLEANFILES = java_bytecode$(LIBEXT) core-models.jar converter
```
Missing `$(EXEEXT)`

cesaro commented on src/java_bytecode/Makefile:
```diff
@@ -36,10 +36,53 @@ INCLUDES= -I ..
+clean : clean_
+
+clean_ :
+	$(RM) -Rf $(BINARY_DIR)
```
Clean `java_core_models.inc` as well, adding it to `CLEANFILES`.

cesaro replied:
Done

tautschnig commented on appveyor.yml:
```diff
@@ -51,6 +51,9 @@ build_script:
     sed -i "s/BUILD_ENV[ ]*=.*/BUILD_ENV = MSVC/" src/config.inc
+    cd src
+    make java_core_models.inc -C java_bytecode -j2
+    cd ..
```
This is caused by using `java_class_loader.o` instead of `java_class_loader$(OBJEXT)` in the Makefile.

cesaro commented on src/java_bytecode/Makefile:
```diff
@@ -36,10 +36,52 @@ INCLUDES= -I ..
+JAR := jar
+JARFLAGS := -cf
+
+.NOTPARALLEL: core-models.jar
```
The use of `.NOTPARALLEL` makes the execution of make on the **entire** folder to be not parallel, which is an unacceptable performance hit. @dcattaruzza will create a new `Makefile` inside of the `library` folder and make that one sequential.

cesaro replied:
done

tautschnig commented on src/java_bytecode/Makefile:
```diff
@@ -36,10 +36,52 @@ INCLUDES= -I ..
 all: java_bytecode$(LIBEXT)
 
+converter: library/converter.cpp
+	$(CXX) $(CP_CXXFLAGS) -o $@ $<
```
```
converter$(EXEEXT): library/converter.cpp
  $(LINKNATIVE)
```

tautschnig commented on src/java_bytecode/Makefile:
```diff
@@ -36,51 +36,17 @@ INCLUDES= -I ..
 
 clean_ :
-	$(RM) -Rf $(BINARY_DIR)
+	make clean -f Makefile1
```
`Makefile1` ? Likely that should be:
```
clean: clean_library

clean_library:
  $(MAKE) -C library clean

.PHONY: clean_library
```

tautschnig commented on src/java_bytecode/Makefile:
```diff
@@ -36,51 +36,17 @@ INCLUDES= -I ..
-
-java_class_loader$(OBJEXT): java_core_models.inc java_class_loader.cpp
+java_class_loader$(OBJEXT): library/java_core_models.inc java_class_loader.cpp
+	make -C library
```
```
library/java_core_models.inc:
  $(MAKE) -C library java_core_models.inc

java_class_loader$(OBJEXT): library/java_core_models.inc
```

thk123 commented:
Can I request a TG pointer bump for this.

cesaro commented on src/java_bytecode/library/Makefile:
```diff
@@ -1,3 +1,46 @@
+JAR := jar
+JARFLAGS := -cf
+
+.NOTPARALLEL: core-models.jar
```
Remove the dependencies of the `.NOTPARALLEL` target and move it to the top of the file. Add a comment explaining that we use it to prevent parallel compilation of the java files and why.

cesaro replied:
done

cesaro commented on src/java_bytecode/CMakeLists.txt:
```diff
@@ -5,20 +5,20 @@ file(GLOB_RECURSE java_sources "*.java")
 add_jar(${JAR_NAME} ${java_sources})
 get_filename_component(CORE_MODELS_JAR "library/core-models.jar" REALPATH BASE_DIR ${CMAKE_CURRENT_SOURCE_DIR})
-get_filename_component(JAVA_CORE_MODELS_INC "java_core_models.inc" REALPATH BASE_DIR ${CMAKE_CURRENT_SOURCE_DIR})
+get_filename_component(JAVA_CORE_MODELS_INC "library/java_core_models.inc" REALPATH BASE_DIR ${CMAKE_CURRENT_SOURCE_DIR})
```
Please merge this commit (the 4th one) with the second commit, as you are modifying the changes introduced there.

cesaro replied:
done

cesaro commented on src/java_bytecode/library/Makefile:
```diff
@@ -1,26 +1,46 @@
+
+all: java_core_models.inc
+
+converter: converter.cpp
```
Write `SRC = converter.cpp` instead of giving a explicit rule to build the converter, as it is done elsewhere in the codebase to compile .cpp files to .o files. Move this line to the top of the file, as in other Makefiles. That way 'make clean' will also remove converter.d, which wasn't being removed.

cesaro replied:
done

tautschnig commented on src/java_bytecode/Makefile:
```diff
@@ -38,8 +38,18 @@ include ../common
 all: java_bytecode$(LIBEXT)
 
+library/java_core_models.inc:
+	$(MAKE) -C library
```
I don't think this will work: the default target in that Makefile is the first one, which is `$(BINARY_DIR)`. I believe you will need to add `java_core_models.inc` to this line here.

dcattaruzza replied:
Isn't the default target
all: java_core_models.inc
?

tautschnig replied:
No, it's the first target in a Makefile (except for those starting with a `.`): https://www.gnu.org/software/make/manual/html_node/Goals.html

Degiorgio commented:
@tautschnig  take another look at this please/

peterschrammel reviewed (changes requested):
The last commit mixes up several changes that are not explained by the commit messages 'fix cmake build'. Please split.

Degiorgio commented:
@peterschrammel  directly including the cpp file was causing  linker issues when building with cmake, hence the  "fix cmake build" (specifically,  ODR violations related with the widespread use of globbing in cmake ) .   Notwithstanding this,  the commit message is lacking  details (will address this).    Question:  since all changes in the commit are  addressing  issues raised in this pr relating the cmake file and the aforementioned  error, would you still like it to be split? or should I simply squash?

peterschrammel commented:
@degiorgio, if the commit that introduced the problem is part of the same PR then you should fix that commit rather than adding another commit.

Degiorgio commented:
@peterschrammel  The PR did not introduce the problem, the problem was  already there, it was  triggered by this PR. There are two ways to fix it (that I can think of).  Avoid using globbing in CMake (which is currently being used  everywhere), or  simply avoid including cpp files directly (as done by this pr)

peterschrammel commented:
@Degiorgio, the last commit is doing two things:
1. fixing a bug relating to the inclusion of the .cpp file
2. Reworking cmake compilation that has been modified in the previous commit.

I suggest to
* squash 2. into the previous commit,
* put 1. into a separate commit.

Degiorgio committed:
Fixed CMake linker ODR violations caused by a regression-test

Including .Cpp file in a regression-test was causing linker issues with CMake,
specifically ODR (One Definition Rule) violations. Commit addresses this by
modifying the test to include the header file instead.

Degiorgio commented:
@peterschrammel done

peterschrammel commented on src/java_bytecode/java_bytecode_language.h:
```diff
@@ -35,19 +36,21 @@ Author: Daniel Kroening, kroening@kroening.com
-  " --lazy-methods-extra-entry-point METHODNAME\n"                                              \
-  "                                  treat METHODNAME as a possible program entry point for\n"  \
-  "                                  the purpose of lazy method loading\n"                      \
+  " --no-core-models                 don't load internally provided models for core classes in\n"/* NOLINT(*) */ \
```
use // clang-format off

Degiorgio replied:
@peterschrammel  just to clarify:   I should replace the /* NOLINT(*) */ with // clang-format off ?  should this be done for all options or just the one added?

peterschrammel commented:
LGTM modulo making clang-format happy

cesaro commented:
@Degiorgio please check that the corresponding TG pointer bump (1406) is correctly pointing to the tip of this PR

tautschnig commented on src/java_bytecode/Makefile:
```diff
@@ -38,8 +38,18 @@ include ../common
 
 CLEANFILES = java_bytecode$(LIBEXT)
 
+clean : clean_library
```
Is there a whitespace after `clean`? I'm not sure what make's semantics of this are, but I'd rather avoid it. The same seems the be true of the `clean_library` goal.

tautschnig commented on src/java_bytecode/Makefile:
```diff
@@ -38,8 +38,18 @@ include ../common
+
+.PHONY: clean_library
+clean_library :
+	$(MAKE) clean -C library
```
Please move all of the above changes below the declaration of the `all` goal so that `all` remains the default goal.

tautschnig commented on src/java_bytecode/java_class_loader.h:
```diff
@@ -68,6 +81,18 @@ class java_class_loadert:public messaget
+  /// or may not exist since  the actual data bis retrieved from memory.
+  jar_filet &jar_pool(java_class_loader_limitt &limit,
+                      const std::string &buffer_name,
+                      const void *pMem,
```
No camelcase, please.

tautschnig commented on src/java_bytecode/library/Makefile:
```diff
@@ -1,26 +1,51 @@
-	$(JAVAC) $(JFLAGS) $(all_javas)
+.SUFFIXES: .java .class
+
+$(BINARY_DIR)/%.class : $(SOURCE_DIR)/%.java $(BINARY_DIR)
```
No space before `:`

Degiorgio replied:
done

tautschnig commented on src/java_bytecode/library/Makefile:
```diff
@@ -1,26 +1,51 @@
+clean_ :
+	$(RM) -Rf $(BINARY_DIR)
+
+all: java_core_models.inc
```
Please make the `all` goal the first one in this file to make it the default.

Degiorgio replied:
done

tautschnig commented on src/java_bytecode/library/Makefile:
```diff
@@ -1,26 +1,51 @@
+all: java_core_models.inc
+
+converter$(EXEEXT): converter.cpp
+	$(CXX) $(CP_CXXFLAGS) -o $@ $<
```
Please use `$(LINKNATIVE)` instead of `$(CXX) ....`

Degiorgio replied:
done

tautschnig commented on src/java_bytecode/library/Makefile:
```diff
@@ -1,26 +1,51 @@
+
+java_core_models.inc: converter$(EXEEXT) core-models.jar
+	./converter$(EXEEXT) JAVA_CORE_MODELS core-models.jar > $@
+###############################################################################
```
I don't think this line is of much value. In other Makefiles, it provides some separation of somewhat "common" rules from very local ones, but using it as the very last line seems pointless.

Degiorgio replied:
done

tautschnig commented on src/java_bytecode/library/converter.cpp:
```diff
@@ -0,0 +1,58 @@
+
+  if(argc<2 || strlen(argv[1])==0)
+  {
+    std::cout << "Usage: converter VARNAME\n";
```
It seems there can also be optional additional arguments? Please document them.

Degiorgio replied:
done

tautschnig commented on src/java_bytecode/library/converter.cpp:
```diff
@@ -0,0 +1,58 @@
+  }
+  varname=argv[1];
+
+  printf("// File automatically generated by " __FILE__ "\n\n");
```
Use `std::cout` consistently, no need for `printf`

Degiorgio replied:
going to use printf consistently instead as it is a simpler change. For compatibility reasons (printing size_t  using printf is problematic in c++98) the last one uses std::cout.

tautschnig commented on src/java_bytecode/library/converter.cpp:
```diff
@@ -0,0 +1,58 @@
+
+  printf("// File automatically generated by " __FILE__ "\n\n");
+
+  size=0;
```
As this is C++ code, please move the declaration to the point of definition, and please make this a `std::size_t` to avoid overflow.

tautschnig commented on src/java_bytecode/library/converter.cpp:
```diff
@@ -0,0 +1,58 @@
+    std::cout << "\n";
+    return 1;
+  }
+  varname=argv[1];
```
C++: declare at the point of definition (and you might consider using `std::string`).

Degiorgio replied:
done

tautschnig commented on src/java_bytecode/library/converter.cpp:
```diff
@@ -0,0 +1,58 @@
+  std::ifstream src((argc>2) ? argv[2] : "core-models.jar",
+    std::ios_base::binary);
+  std::istream &stream=(argc>2) ? src : std::cin;
+  if(!stream.eof())
```
You might want to check `.good`

Degiorgio commented:
@tautschnig addressed issues you raised, can you take another look please?

cesaro commented:
@tautschnig All the issues you raised seem to have been addressed, please have a look, we want to merge in a couple of hours, it's blocking work on another PR

cesaro commented:
The TG pointer update is passing (1405)

cesaro commented:
@tautschnig After verbally discussing with Peter we decide to dismiss your review from 10 days ago, subsumed by a more recent one also by you 3 days ago, as all of the changes have been implemented and there are already 2 approvals.

tautschnig commented:
> @tautschnig After verbally discussing with Peter we decide to dismiss your review from 10 days ago [...]

My apologies for having huge latency on my part, thanks a lot for taking the appropriate action!

cesaro commented:
@tautschnig Many thanks to you for the suggestions for improvement that you offered!
